scipy/sparse/csgraph/_laplacian.py

```diff
@@ -37,118 +41,510 @@ def laplacian(csgraph, normed=False, return_diag=False, use_out_degree=False,
+    must be used in the construction of the Laplacian.
+    The option ``normed=True`` cannot be used for the negative weights here
+    as the symmetric normalization evaluates square roots.
+    The option ``form="lo"`` is matrix-free, i.e., guarantees
+    a fixed memory footprint and read-only access to the graph.
+
+    ...     L = csgraph.laplacian(G, symmetrized=True, form="lo")
+
+    Calling the eigenvalue solver ``lobpcg`` to compute the Fiedler vector
+    that determines the labels as the signs of its components.
+
+    ...     _, eves = sparse.linalg.lobpcg(L, X, Y=Y, largest=False, tol=1e-3)
+    ...     print(cut + "-cut labels:\\n", 1 * (eves[:, 0]>0))
+    max-cut labels:
+    [0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0]
+    min-cut labels:
+    [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1]
+
+    As anticipated for a (slightly noisy) linear graph,
+    the max-cut strips all the edges of the graph coloring all
+    odd vertices into one color and all even vertices into another one,
+    while the balanced min-cut partitions the graph
+    in the middle by deleting a single edge.
+    Both determined partitions are optimal.
     """
     if csgraph.ndim != 2 or csgraph.shape[0] != csgraph.shape[1]:
-        raise ValueError('csgraph must be a square matrix or array')
+        raise ValueError("csgraph must be a square matrix or array")
 
-    if normed and (np.issubdtype(csgraph.dtype, np.signedinteger)
-                   or np.issubdtype(csgraph.dtype, np.uint)):
-        csgraph = csgraph.astype(float)
+    if normed and (
+        np.issubdtype(csgraph.dtype, np.signedinteger)
+        or np.issubdtype(csgraph.dtype, np.uint)
+    ):
+        csgraph = csgraph.astype(np.float64)
+
+    if form == "array":
+        create_lap = (
+            _laplacian_sparse if isspmatrix(csgraph) else _laplacian_dense
+        )
+    else:
+        create_lap = (
+            _laplacian_sparse_flo
+            if isspmatrix(csgraph)
+            else _laplacian_dense_flo
+        )
 
-    create_lap = _laplacian_sparse if isspmatrix(csgraph) else _laplacian_dense
     degree_axis = 1 if use_out_degree else 0
-    lap, d = create_lap(csgraph, normed=normed, axis=degree_axis,
-                        copy=copy)
+
+    lap, d = create_lap(
+        csgraph,
+        normed=normed,
+        axis=degree_axis,
+        copy=copy,
+        form=form,
+        dtype=dtype,
+        symmetrized=symmetrized,
+    )
     if return_diag:
         return lap, d
     return lap
 
 
-def _setdiag_dense(A, d):
-    A.flat[::len(d)+1] = d
+def _setdiag_dense(m, d):
+    step = len(d) + 1
+    m.flat[::step] = d
+
+
+def _laplace(m, d):
+    return lambda v: v * d[:, np.newaxis] - m @ v
+
+
+def _laplace_normed(m, d, nd):
+    laplace = _laplace(m, d)
+    return (
+        lambda v: nd[:, np.newaxis] * laplace(v * nd[:, np.newaxis])
+           )
```

Comment:
Never seen such formatting
```suggestion
    return (
        lambda v: nd[:, np.newaxis] * laplace(v * nd[:, np.newaxis])
    )
```